hazelcast/src/main/java/com/hazelcast/sql/impl/expression/string/LikeFunction.java

```diff
@@ -0,0 +1,214 @@
+
+        String pattern = asVarchar(operand2, row, context);
+
+        if (pattern == null) {
+            return null;
+        }
+
+        String escape;
+
+        if (operand3 != null) {
+            escape = asVarchar(operand3, row, context);
+
+            if (escape == null) {
+                return null;
+            }
+        } else {
+            escape = null;
+        }
+
+        if (state == null) {
+            state = new State();
+        }
+
+        return state.like(source, pattern, escape);
+    }
+
+    @Override
+    public QueryDataType getType() {
+        return QueryDataType.BOOLEAN;
+    }
+
+    @Override
+    public int getFactoryId() {
+        return SqlDataSerializerHook.F_ID;
+    }
+
+    @Override
+    public int getClassId() {
+        return SqlDataSerializerHook.EXPRESSION_LIKE;
+    }
+
+    /**
+     * Helper class to execute LIKE function. Caches the last observed pattern to avoid constant re-compilation.
+     */
+    public static class State {
+        /** Last observed pattern. */
+        private String lastPattern;
+
+        /** Last observed escape. */
+        private String lastEscape;
+
+        /** Last Java pattern. */
+        private Pattern lastJavaPattern;
+
+        public boolean like(String source, String pattern, String escape) {
+            Pattern javaPattern = convertToJavaPattern(pattern, escape);
+
+            Matcher matcher = javaPattern.matcher(source);
+
+            return matcher.matches();
+        }
+
+        private Pattern convertToJavaPattern(String pattern, String escape) {
+            if (lastJavaPattern != null && Objects.equals(pattern, lastPattern) && Objects.equals(escape, lastEscape)) {
+                return lastJavaPattern;
+            }
+
+            String javaPatternStr = constructJavaPatternString(pattern, escape);
+            Pattern javaPattern = Pattern.compile(javaPatternStr);
+
+            lastPattern = pattern;
+            lastEscape = escape;
+            lastJavaPattern = javaPattern;
+
+            return javaPattern;
+        }
+
+        @SuppressWarnings("checkstyle:CyclomaticComplexity")
+        private static String constructJavaPatternString(String pattern, String escape) {
+            // Get the escape character.
+            Character escapeChar;
+
+            if (escape != null) {
+                if (escape.length() != 1) {
+                    throw QueryException.error("Escape parameter should be a single character: " + escape);
+                }
+
+                escapeChar = escape.charAt(0);
+            } else {
+                escapeChar = null;
+            }
+
+            // Main logic.
+            StringBuilder javaPattern = new StringBuilder();
+
+            int i;
+
+            for (i = 0; i < pattern.length(); i++) {
+                char patternChar = pattern.charAt(i);
+
+                // Escape special character as needed.
+                if (ESCAPE_CHARACTERS_JAVA.indexOf(patternChar) >= 0) {
```

Comment:
If `pattern = "(_"` and `escape = "("`, we'll produce the regex `\_` - the non-special `_` will be unnecessarily escaped. This happens when the escape character is one of the regex special characters.

While I couldn't find any case where this is an issue (in regex you can unnecessarily escape any character), seems weird to produce such output.

Reply:
I am not quite sure I understand the problem. Does it lead to incorrect results in any case?

Reply:
It only causes that `_` or `%` can be unnecessarily escaped. But since regex allows escaping any non-alphanumerical characters, it's not really a problem.